xiilib/_gunicorn/charm.py

```diff
@@ -0,0 +1,191 @@
+from xiilib.exceptions import CharmConfigInvalidError
+
+logger = logging.getLogger(__name__)
+
+
+class GunicornBase(abc.ABC, ops.CharmBase):  # pylint: disable=too-many-instance-attributes
+    """Gunicorn-based charm service mixin."""
+
+    @abc.abstractmethod
+    def get_wsgi_config(self) -> BaseModel:
+        """Return the framework related configurations."""
+
+    @abc.abstractmethod
+    def get_cos_dir(self) -> str:
+        """Return the directory with COS related files."""
+
+    def __init__(self, framework: ops.Framework, wsgi_framework: str) -> None:
+        """Initialize the instance.
+
+        Args:
+            framework: operator framework.
+            wsgi_framework: WSGI framework name.
+        """
+        super().__init__(framework)
+        self._secret_storage = GunicornSecretStorage(
+            charm=self, key=f"{wsgi_framework}_secret_key"
+        )
+        self._database_requirers = make_database_requirers(self, self.app.name)
+        try:
+            wsgi_config = self.get_wsgi_config()
+        except CharmConfigInvalidError as exc:
+            self._update_app_and_unit_status(ops.BlockedStatus(exc.msg))
+            return
+
+        self._charm_state = CharmState.from_charm(
+            charm=self,
+            framework=wsgi_framework,
+            wsgi_config=wsgi_config,
+            secret_storage=self._secret_storage,
+            database_requirers=self._database_requirers,
+        )
+        self._database_migration = DatabaseMigration(
+            container=self.unit.get_container(self._charm_state.container_name),
+            state_dir=self._charm_state.state_dir,
+        )
+        webserver = GunicornWebserver(
+            charm_state=self._charm_state,
+            container=self.unit.get_container(self._charm_state.container_name),
+        )
+        self._wsgi_app = WsgiApp(
+            container=self.unit.get_container(f"{self._charm_state.framework}-app"),
+            charm_state=self._charm_state,
+            webserver=webserver,
+            database_migration=self._database_migration,
+        )
+        self._databases = Databases(
+            charm=self,
+            application=self._wsgi_app,
+            database_requirers=self._database_requirers,
+        )
+        self._ingress = IngressPerAppRequirer(
+            self,
+            port=self._charm_state.port,
+            strip_prefix=True,
+        )
+        self._observability = Observability(
+            self,
+            charm_state=self._charm_state,
+            container_name=self._charm_state.container_name,
+            cos_dir=self.get_cos_dir(),
+        )
+        self.framework.observe(self.on.config_changed, self._on_config_changed)
+        self.framework.observe(self.on.rotate_secret_key_action, self._on_rotate_secret_key_action)
+        self.framework.observe(
+            self.on.secret_storage_relation_changed,
+            self._on_secret_storage_relation_changed,
+        )
```

Comment:
Why don't we use our observer pattern here and observe from the `secret_storage` class ?

Reply:
I think it will have circular dependency like databases

Reply:
The hook method just call restart which is then delegated to wsgi_app, am I missing something ?

Reply:
`wsgi_app` needs `charm_state`, `charm_state` needs `secret_storage` for secret key, if the `secret_storage` is an observer and needs `wsgi_app` to restart services, then it just forms a dependency triangle.